client/cli/src/commands/mod.rs

```diff
@@ -84,56 +84,70 @@ impl Subcommand {
 		}
 	}
 
-	/// Run any `CoreParams` command.
-	pub fn run<B, BC, BB>(
-		self,
-		config: Configuration,
-		builder: B,
-	) -> error::Result<()>
-	where
-		B: FnOnce(Configuration) -> Result<BC, sc_service::error::Error>,
-		BC: ServiceBuilderCommand<Block = BB> + Unpin,
-		BB: sp_runtime::traits::Block + Debug,
-		<<<BB as BlockT>::Header as HeaderT>::Number as std::str::FromStr>::Err: std::fmt::Debug,
-		<BB as BlockT>::Hash: std::str::FromStr,
-	{
-		match self {
-			Subcommand::BuildSpec(cmd) => cmd.run(config),
-			Subcommand::ExportBlocks(cmd) => cmd.run(config, builder),
-			Subcommand::ImportBlocks(cmd) => cmd.run(config, builder),
-			Subcommand::CheckBlock(cmd) => cmd.run(config, builder),
-			Subcommand::PurgeChain(cmd) => cmd.run(config),
-			Subcommand::Revert(cmd) => cmd.run(config, builder),
-		}
-	}
-
-	/// Update and prepare a `Configuration` with command line parameters.
-	pub fn update_config<F>(
-		&self,
-		mut config: &mut Configuration,
-		spec_factory: F,
-		version: &VersionInfo,
-	) -> error::Result<()> where
-		F: FnOnce(&str) -> Result<Box<dyn ChainSpec>, String>,
-	{
-		match self {
-			Subcommand::BuildSpec(cmd) => cmd.update_config(&mut config, spec_factory, version),
-			Subcommand::ExportBlocks(cmd) => cmd.update_config(&mut config, spec_factory, version),
-			Subcommand::ImportBlocks(cmd) => cmd.update_config(&mut config, spec_factory, version),
-			Subcommand::CheckBlock(cmd) => cmd.update_config(&mut config, spec_factory, version),
-			Subcommand::PurgeChain(cmd) => cmd.update_config(&mut config, spec_factory, version),
-			Subcommand::Revert(cmd) => cmd.update_config(&mut config, spec_factory, version),
-		}
-	}
-
 	/// Initialize substrate. This must be done only once.
 	///
 	/// This method:
 	///
 	/// 1. Set the panic handler
 	/// 2. Raise the FD limit
 	/// 3. Initialize the logger
-	pub fn init(&self, version: &VersionInfo) -> error::Result<()> {
-		self.get_shared_params().init(version)
+	pub fn init<C: SubstrateCLI>(&self) -> Result<()> {
+		self.get_shared_params().init::<C>()
 	}
 }
+
+macro_rules! match_and_call {
```

Comment:
You're trying to combine a big list of subcommands with a big list of methods. But instead of putting the subcommands in the macro, I would have put the methods.

In other words, something like that:

```rust
macro_rules! impl_cli_config {
    ($struct:ident, $($variants:ident)*) => {
        impl CliConfiguration for $struct {
            fn base_path(&self) -> Result<Option<&PathBuf>> {
                match self { $($struct::$variants(cmd) => { cmd.base_path() })* }
            }

            // ... same for the other methods ...
        }
    }
}

impl_cli_config!(Subcommand, BuildSpec, ExportBlocks, ImportBlocks, CheckBlock, Revert, PurgeChain);
```

Reply:
Fair point!

Reply:
It looks like we could also do a derive macro but then it would have the same problems that I mentioned there: https://github.com/paritytech/substrate/pull/5271#discussion_r397117504